### ISSUE TITLE
Contributing

### ISSUE THREAD

**Issue by SketchNI** [fourum/fourum#1]
I saw your comment on [reddit](https://www.reddit.com/r/PHP/comments/2rlz0w/ask_rphp_open_source_php_based_forums_whats_the/) and I'd like to chip in. What is your goal for first release and what are your contrib guidelines?

**Comment by wpillar** [comment]
@SketchNI awesome, I'll write some stuff down tonight that might help.

**Comment by SketchNI** [comment]
@wpillar thanks!

**Comment by wpillar** [comment]
@SketchNI hey, I've jotted some stuff down here to consider about contributing. Hopefully it gives you enough of an idea to get started!

Firstly some things to know:

- It's built using Laravel 4.1 (I will update to L5 when it launches but we don't need to worry about that for now).
- PSR standards apply, PSR-1 and PSR-2.
- As of now it's a functional forum, it's just not very pretty or feature-rich.
- To set it up, there's a Vagrantfile, clone the repo, setup vagrant, then run `make install` to setup the DB and populate some test data.
- There is very little documentation so far.

As for what to contribute, I'm building this in a way that makes plugins really easy using events, so if you look at https://github.com/fourum/fourum/tree/master/workbench/fourum there's two laravel packages here for two parts of the system, messaging and warnings. They are decent examples of how to write packages for Fourum. See here for more: http://laravel.com/docs/4.1/packages

So with that in mind, here's some packages I think Fourum needs, feel free to do one of those or if you think of something cool, then do that :)

- Moderation (posts, threads etc)
   - Edit posts
   - Delete posts
   - Move threads
   - Delete threads
   - Control Panel
- Banning
   - Arbitrary ban lengths (hours, days, months, years)
   - Choose Rule(s) which were broken
   - Appeals ?
   - Banned user group ?
   - Ban voting ?
   - Cookie banning to prevent multiple accounts on the same computer

Hopefully that's enough for you to get involved, don't worry if this is too much and you don't fancy it! Any questions, you can drop me an email (me at willpillar dot com).

**Comment by SketchNI** [comment]
Hey, thanks for getting back to me. I'll review the packages and get stuck in tomorrow!

**Comment by splitfeed** [comment]
I got here through the reddit thread as well and wanted to check it out, but I can't even seem to get past composer. 

    Script php artisan optimize handling the post-install-cmd event returned with an error
    [RuntimeException]
    Error Output: PHP Fatal error:  Class 'Fourum\Message\MessageServiceProvider' not found

I haven't been working with the workbench much so it might just be my mistake, but in that case I'd love for some pointer on how to get it running :)

**Comment by SketchNI** [comment]
@splitfeed what exact steps are you taking? (also out of curiosity, what OS and version are you using because you may have a little trouble with vagrant as well)

**Comment by wpillar** [comment]
@SketchNI did you get it up and running ok?

**Comment by SketchNI** [comment]
No, I'm getting the same error as @splitfeed when running ````make install````

**Comment by splitfeed** [comment]
Well, I never went past cloning and then running make install, or just composer install.

I am running it on my Debian server (as Vagrant on my Windows desktop is a pain) . Vagrant will hopefully not be an issue as I have other Laravel projects using Homestead :)

**Comment by wpillar** [comment]
Try `php artisan dump-autoload` running it through the artisan wrapper includes the workbench packages.

**Comment by splitfeed** [comment]
Same error, seems something is amiss with loading from the workbench. If I reorder the providers in app.php the same error goes out for the warning package as well.

$ php artisan dump-autoload
PHP Fatal error:  Class 'Fourum\Message\MessageServiceProvider' not found in /code/Fourum/vendor/laravel/framework/src/Illuminate/Foundation/ProviderRepository.php on line 123
PHP Stack trace:
PHP   1. {main}() /code/Fourum/artisan:0
PHP   2. require_once() /code/Fourum/artisan:30
PHP   3. require() /code/Fourum/bootstrap/start.php:57
PHP   4. Illuminate\Foundation\ProviderRepository->load() /code/Fourum/vendor/laravel/framework/src/Illuminate/Foundation/start.php:183
PHP   5. Illuminate\Foundation\ProviderRepository->compileManifest() /code/Fourum/vendor/laravel/framework/src/Illuminate/Foundation/ProviderRepository.php:51
PHP   6. Illuminate\Foundation\ProviderRepository->createProvider() /code/Fourum/vendor/laravel/framework/src/Illuminate/Foundation/ProviderRepository.php:89

Can you reproduce this yourself with a clean clone?

**Comment by SketchNI** [comment]
````
$ php artisan dump-autoload
PHP Fatal error:  Class 'Fourum\Message\MessageServiceProvider' not found in /srv/fourum/vendor/laravel/framework/src/Illuminate/Foundation/ProviderRepository.php on line 123
{"error":{"type":"Symfony\\Component\\Debug\\Exception\\FatalErrorException","message":"Class 'Fourum\\Message\\MessageServiceProvider' not found","file":"\/srv\/fourum\/vendor\/laravel\/framework\/src\/Illuminate\/Foundation\/ProviderRepository.php","line":123}}{"error":{"type":"Symfony\\Component\\Debug\\Exception\\FatalErrorException","message":"Class 'Fourum\\Message\\MessageServiceProvider' not found","file":"\/srv\/fourum\/vendor\/laravel\/framework\/src\/Illuminate\/Foundation\/ProviderRepository.php","line":123}}
````

I got the same error, only with JSON output to terminal on Ubuntu MATE Edition.

**Comment by wpillar** [comment]
Ok, I'll try and reproduce when I get home tonight. Sorry!

**Comment by wpillar** [comment]
Ok, got something that worked for me, try these steps:

1. Pull on master.
2. Open `fourum/config/app.php`.
3. Comment out `Fourum\Message\MessageServiceProvider`.
4. Run `php artisan dump-autoload` (this should now pass and you should see it doing it for each package).
5. If the above executes successfully, open up `fourum/config/app.php` again and add `Fourum\Message\MessageServiceProvider` back.
6. You should now have a working application. You may need to run `make install` again first.

Let me know how that works out for you guys, apparently this is a common workbench problem in Laravel.

**Comment by splitfeed** [comment]
That did it for me, thanks! :)

**Comment by SketchNI** [comment]
Removed fourum completely, cloned a clean copy, tried this steps, still get an error about Fourum\Message\MessageServiceProvider but I'm now getting an error about Fourum\Warning\WarningServiceProvider.

**Comment by wpillar** [comment]
Do the same steps for that one too, so get rid of both `Fourum\Message\MessageServiceProvider` and `Fourum\Warning\WarningServiceProvider`. Then do `php artisan dump-autoload` and then put them back.
